### PR TITLE
[WIP] Define and implement new cudf::shift APIs

### PR DESCRIPTION
**Pull request by cwharris**
This PR implements functionality required to complete #2434, #2390, and #1285, but does not address any python APIs.

### CODE REVIEW

**Comment by cwharris** [comment]
This API is written for columns and not tables because it allows for a `fill_value`, which will differ depending on column type.

**Comment by harrism** [comment]
> Also includes some tabs => spaces housekeeping.
> 
> #2434
> #2390
> #1285

If this PR fixes those issues, please write `fixes  #<issue>` so they get autoclosed when this merges.

**Comment by harrism** [comment]
> This API is written for columns and not tables because it allows for a `fill_value`, which will differ depending on column type.

You can shift tables by providing `std::vector<cudf::scalar> fill_values`.

File: cpp/include/cudf/copying.hpp [hunk 1]
```diff
@@ -173,5 +173,27 @@ std::vector<column_view> slice(column_view const& input,
 std::vector<column_view> split(column_view const& input,
                                std::vector<size_type> const& splits);
 
+/**
+* @brief Shifts a `column_view` by `periods`, using a fill_value for
+* pre- or post-fix data.
+*
+* Output will be nullable if either the input is nullable or the fill value is null.
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp]:
```suggestion
* The output is nullable if either the input is nullable or the fill value is null.
```

File: cpp/include/cudf/copying.hpp [hunk 2]
```diff
@@ -173,5 +173,27 @@ std::vector<column_view> slice(column_view const& input,
...
+*
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp]: Document how direction of shift is determined (positive/negative periods?)

File: cpp/src/copying/shift.cu [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+ #include <cudf/copying.hpp>
+ #include <cudf/table/table.hpp>
+ #include <cudf/table/table_view.hpp>
+ #include <cudf/scalar/scalar.hpp>
+
+ namespace cudf {
+
+ std::unique_ptr<table> shift(const table_view& in,
```
Comment (cwharris) [inline cpp/src/copying/shift.cu]: Just realized these should be `_ const&`, not `const _ &`. Still getting used to c++. :)

Comment (harrism) [reply cpp/src/copying/shift.cu]: East const is only a matter of preference. But I'm getting into the habit.

Comment (cwharris) [reply cpp/src/copying/shift.cu]: I was wondering why I didn't see any issues pop up. Still reading my C++ books, and enjoying it thoroughly.

File: cpp/include/cudf/copying.hpp [hunk 4]
```diff
@@ -173,5 +174,31 @@ std::vector<column_view> slice(column_view const& input,
...
+* column offset by `periods`, using `fill_value` when input is out-of-bounds.
+*
+* When `i + periods` is in bounds:
+*   `out[col][i] = in[col][i + periods]`
+*
+* When `i + periods` is out of bounds and `fill_value` is supplied:
+*   `out[col][i] = fill_value[i]`
+*
+* When `i + periods` is out of bounds and `fill_value is not supplied:
+*   `out[col][i] = null`
```
Comment (devavret) [inline cpp/include/cudf/copying.hpp]: Examples are much better at explaining what's happening.

File: cpp/include/cudf/copying.hpp [hunk 5]
```diff
@@ -173,5 +174,31 @@ std::vector<column_view> slice(column_view const& input,
...
+*
+* An output column is nullable if either the corrosponding input column or
```
Comment (devavret) [inline cpp/include/cudf/copying.hpp]:
```suggestion
* An output column is nullable if either the corresponding input column or
```

File: cpp/include/cudf/copying.hpp [hunk 6]
```diff
@@ -173,5 +174,31 @@ std::vector<column_view> slice(column_view const& input,
 std::vector<column_view> split(column_view const& input,
                                std::vector<size_type> const& splits);
 
+/**
+* @brief Creates a new `table` similar to the input but with values in every
+* column offset by `periods`, using `fill_value` when input is out-of-bounds.
+*
+* When `i + periods` is in bounds:
+*   `out[col][i] = in[col][i + periods]`
```
Comment (devavret) [inline cpp/include/cudf/copying.hpp]: So this is the reverse of pandas? If out[i] == in[i + period] then for in = [1,2,3,4,5], period = 2, out = [3,4,5,None,None]? But in Pandas, the shift goes in the other way

File: cpp/src/copying/shift.cu [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+std::unique_ptr<table> shift(table_view const& in,
+                              size_type periods,
+                              std::vector<scalar> const& fill_values,
+                              rmm::mr::device_memory_resource *mr)
+{
+    if (not fill_values.empty()) {
+        CUDF_EXPECTS(static_cast <unsigned int>(in.num_columns()) == fill_values.size(), "`fill_values.size()` and `in.num_columns() must be the same.");
+    }
+
+    if (periods == 0 || in.num_rows() == 0) {
```
Comment (devavret) [inline cpp/src/copying/shift.cu]: if periods == 0, wouldn't the result be the same as input?

Comment (cwharris) [reply cpp/src/copying/shift.cu]: Yes.

**Comment by harrism** [comment]
Moving to 0.12.

File: cpp/include/cudf/copying.hpp (hunk 6 above)
Comment (cwharris) [reply cpp/include/cudf/copying.hpp]: Addressing these concerns in #3662.

File: cpp/include/cudf/copying.hpp (hunk 5 above)
Comment (cwharris) [reply cpp/include/cudf/copying.hpp]: Addressing these concerns in #3662.

File: cpp/include/cudf/copying.hpp (hunk 4 above)
Comment (cwharris) [reply cpp/include/cudf/copying.hpp]: Addressing these concerns in #3662.

File: cpp/src/copying/shift.cu (hunk 7 above)
Comment (cwharris) [reply cpp/src/copying/shift.cu]: Addressing these concerns in #3662.

**Pull request closed without merging**